sql/core/benchmarks/DataSourceReadBenchmark-jdk11-results.txt

```diff
@@ -2,322 +2,322 @@
+SQL Parquet Vectorized: DataPageV2                 1939           1941           3          5.4         184.9       3.5X
+SQL Parquet MR: DataPageV1                         3090           3099          13          3.4         294.7       2.2X
+SQL Parquet MR: DataPageV2                         3274           3286          17          3.2         312.3       2.1X
+ParquetReader Vectorized: DataPageV1               1058           1067          13          9.9         100.9       6.4X
+ParquetReader Vectorized: DataPageV2               1847           1848           2          5.7         176.2       3.7X
+SQL ORC Vectorized                                 1307           1307           0          8.0         124.6       5.2X
+SQL ORC MR                                         3078           3122          62          3.4         293.6       2.2X
+
+OpenJDK 64-Bit Server VM 11.0.14+9-LTS on Linux 5.11.0-1025-azure
 Intel(R) Xeon(R) Platinum 8272CL CPU @ 2.60GHz
 String with Nulls Scan (95.0%):           Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
 ------------------------------------------------------------------------------------------------------------------------
-SQL CSV                                            4443           4504          86          2.4         423.7       1.0X
-SQL Json                                           4528           4563          49          2.3         431.8       1.0X
-SQL Parquet Vectorized: DataPageV1                  213            233          15         49.2          20.3      20.8X
-SQL Parquet Vectorized: DataPageV2                  267            294          22         39.3          25.4      16.7X
-SQL Parquet MR: DataPageV1                         1691           1700          13          6.2         161.2       2.6X
-SQL Parquet MR: DataPageV2                         1515           1565          70          6.9         144.5       2.9X
-ParquetReader Vectorized: DataPageV1                228            231           2         46.0          21.7      19.5X
-ParquetReader Vectorized: DataPageV2                285            296           9         36.8          27.1      15.6X
-SQL ORC Vectorized                                  369            425          82         28.4          35.2      12.1X
-SQL ORC MR                                         1457           1463           9          7.2         138.9       3.0X
+SQL CSV                                            4818           4824           8          2.2         459.5       1.0X
+SQL Json                                           4853           4878          35          2.2         462.8       1.0X
+SQL Parquet Vectorized: DataPageV1                  255            264           6         41.1          24.3      18.9X
+SQL Parquet Vectorized: DataPageV2                  711            716           4         14.7          67.8       6.8X
```

Comment:
Hmm I wonder why performance dropped so much.

Reply:
Seems it is related to the percentage of nulls. BTW do you see any benefit from switching to new implementation for `DELTA_BYTE_ARRAY`? The old implementation just relies on `parquet-mr` and is very simple.

Reply:
So I removed the `if (total == 0)` check and it seems to have a big impact on performance. Before:
```
[info] String with Nulls Scan (95.0%):           Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] SQL CSV                                            3155           3213          83          3.3         300.9       1.0X
[info] SQL Json                                           3934           3982          68          2.7         375.2       0.8X
[info] SQL Parquet Vectorized: DataPageV1                  149            188          83         70.3          14.2      21.2X
[info] SQL Parquet Vectorized: DataPageV2                  198            235          55         53.0          18.9      15.9X
[info] SQL Parquet MR: DataPageV1                         1109           1155          66          9.5         105.7       2.8X
[info] SQL Parquet MR: DataPageV2                         1086           1094          12          9.7         103.5       2.9X
[info] ParquetReader Vectorized: DataPageV1                110            112           1         94.9          10.5      28.6X
[info] ParquetReader Vectorized: DataPageV2                444            547          90         23.6          42.3       7.1X
[info] SQL ORC Vectorized                                  960           1008          64         10.9          91.6       3.3X
[info] SQL ORC MR                                         1374           1403          41          7.6         131.0       2.3X
```

After:
```
[info] String with Nulls Scan (95.0%):           Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] SQL CSV                                            3719           4191         667          2.8         354.7       1.0X
[info] SQL Json                                           4774           7948        1303          2.2         455.3       0.8X
[info] SQL Parquet Vectorized: DataPageV1                  172            268         143         61.0          16.4      21.6X
[info] SQL Parquet Vectorized: DataPageV2                  218            302         121         48.2          20.8      17.1X
[info] SQL Parquet MR: DataPageV1                         1030           1075          64         10.2          98.2       3.6X
[info] SQL Parquet MR: DataPageV2                         1919           2783        1222          5.5         183.0       1.9X
[info] ParquetReader Vectorized: DataPageV1                123            228         149         85.3          11.7      30.2X
[info] ParquetReader Vectorized: DataPageV2                169            176           8         62.2          16.1      22.0X
[info] SQL ORC Vectorized                                  269            284           9         39.0          25.7      13.8X
[info] SQL ORC MR                                         1006           1016          14         10.4          96.0       3.7X
```

Reply:
Amazing !!! This is a case worth analyzing

Reply:
After [406d176](https://github.com/apache/spark/pull/35262/commits/406d1768019eed1516925847a59bd2ad5fc7883b), should we update bench result?